src/profilers/profile-serializer.ts

```diff
@@ -15,132 +15,279 @@
+ * when serializing a profile.
+ */
+class StringTable {
+  strings: string[];
+  stringsMap: Map<string, number>;
+
+  constructor() {
+    this.strings = [];
+    this.stringsMap = new Map<string, number>();
+    this.getIndexOrAdd('');
+  }
 
   /**
-   * Adds samples from a node and it's children to the fields tracking
-   * profile serialization.
-   *
-   * @param node - the node which is serialized
-   * @param stack - the stack trace to the current node.
+   * Returns index of str within the table. Also adds str to string table if
+   * str is not in the table already.
    */
-  function serializeNode(root: TimeProfileNode) {
-    let entries: Entry[] = [];
-    // don't include root node in serialized profile, start with it's children.
-    for (let child of root.children) {
-      entries.push({node: child, stack: []});
+  getIndexOrAdd(str: string): number {
+    let loc = this.stringsMap.get(str);
+    if (loc !== undefined) {
+      return loc;
     }
-    while (entries.length > 0) {
-      let entry = entries.pop();
-      if (entry !== undefined) {
-        let node = entry.node;
-        let stack = entry.stack;
-        let location = getLocation(node);
-        stack.unshift(location.id as number);
-        if (node.hitCount > 0) {
-          const sample = new perftools.profiles.Sample({
-            locationId: stack,
-            value: [node.hitCount, node.hitCount * intervalMicros]
-          });
-          samples.push(sample);
-        }
-        for (let child of node.children) {
-          entries.push({node: child, stack: stack.slice(0)});
-        }
+    loc = this.strings.length;
+    this.stringsMap.set(str, loc);
+    this.strings.push(str);
+    return loc;
+  }
+}
+
+/**
+ * Takes v8 profile and populates sample, location, and function fields of
+ * profile.proto.
+ *
+ * @param profile - profile.proto with empty sample, location, and function
+ * fields.
+ * @param root - root of v8 profile tree describing samples to be appended
+ * to profile.
+ * @param appendToSample - function which converts entry to sample(s)  and
+ * appends these to end of an array of samples.
+ * @param stringTable - string table for the existing profile.
+ */
+function serialize(
+    profile: perftools.profiles.IProfile, root: ProfileNode,
+    appendToSample: AppendEntryToSamples, stringTable: StringTable) {
+  const samples: Array<perftools.profiles.Sample> = [];
+  const locations: Array<perftools.profiles.Location> = [];
+  const functions: Array<perftools.profiles.Function> = [];
+  const locationMap: Map<number, perftools.profiles.Location> = new Map();
+  const functionMap: Map<number, perftools.profiles.Function> = new Map();
+  const functionIdMap = new Map<string, number>();
+  const locationIdMap = new Map<string, number>();
+
+  const entries: Entry[] = root.children.map((n) => ({node: n, stack: []}));
+  while (entries.length > 0) {
+    const entry = entries.pop();
+    if (entry !== undefined) {
```

Comment:
Tangential: What's the case where `entry` would be `undefined`?

Reply:
entry should never be undefined, but if I remove that if statement, then I get the error 
"[ts] Object is possibly 'undefined'.
const entry: Entry | undefined"

Is there a more standard way to deal with that?

Reply:
Given that you're checking for length as the loop condition, entry can never be undefined, but the compiler doesn't know that constraint. You can either use a cast or use (type guards)[https://www.typescriptlang.org/docs/handbook/advanced-types.html].

Reply:
@ofrobots Would "const entry = entries.pop()!;" work?

Reply:
@nolanmar511 Did you try what I suggested?

Reply:
Just did -- that works.